File: backend/firebase.py
import os
from dotenv import load_dotenv
import firebase_admin
from firebase_admin import credentials, auth, db, storage

# Load environment variables
load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))

# Safely resolve the service account path
raw_cred_path = os.environ.get('FIREBASE_SERVICE_ACCOUNT_PATH', 'backend/firebase-service-account.json')
if not os.path.isabs(raw_cred_path):
    cred_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', raw_cred_path))
else:
    cred_path = raw_cred_path

if not os.path.exists(cred_path):
    raise FileNotFoundError(f"Firebase credential file not found at: {cred_path}")

# Load other Firebase environment variables
database_url = os.environ.get('FIREBASE_DATABASE_URL')
storage_bucket = os.environ.get('FIREBASE_STORAGE_BUCKET')
project_id = os.environ.get('FIREBASE_PROJECT_ID')

# Initialize Firebase
cred = credentials.Certificate(cred_path)
if not firebase_admin._apps:
    firebase_admin.initialize_app(cred, {
        'databaseURL': database_url,
        'storageBucket': storage_bucket,
        'projectId': project_id
    })

print("✅ Firebase initialized")
print("Service account path:", cred_path)

File: backend-go/firebase/firebase.go
package firebase

import (
	"context"
	"log"
	"os"
	"path/filepath"

	firebase "firebase.google.com/go/v4"
	"firebase.google.com/go/v4/auth"
	"firebase.google.com/go/v4/db"
	"google.golang.org/api/option"

	"echobasket/config"
)

var (
	App        *firebase.App
	AuthClient *auth.Client
	DBClient   *db.Client
)

func Initialize() error {
	ctx := context.Background()

	// Resolve service account path
	credPath := config.AppConfig.FirebaseServiceAccountPath

	// Try multiple locations to find the credential file
	possiblePaths := []string{
		credPath,                           // As specified
		filepath.Join("..", credPath),      // Relative to parent
		"firebase-service-account.json",    // Current directory
		filepath.Base(credPath),            // Just the filename
	}

	foundPath := ""
	for _, p := range possiblePaths {
		if _, err := os.Stat(p); err == nil {
			foundPath = p
			break
		}
	}

	if foundPath == "" {
		log.Fatalf("Firebase credential file not found. Tried: %v", possiblePaths)
	}
	credPath = foundPath

	opt := option.WithCredentialsFile(credPath)

	conf := &firebase.Config{
		DatabaseURL:   config.AppConfig.FirebaseDatabaseURL,
		StorageBucket: config.AppConfig.FirebaseStorageBucket,
		ProjectID:     config.AppConfig.FirebaseProjectID,
	}

	var err error
	App, err = firebase.NewApp(ctx, conf, opt)
	if err != nil {
		return err
	}

	AuthClient, err = App.Auth(ctx)
	if err != nil {
		return err
	}

	DBClient, err = App.Database(ctx)
	if err != nil {
		return err
	}

	log.Println("✅ Firebase initialized")
	log.Printf("Service account path: %s", credPath)

	return nil
}
